Function expressions (#82)
javascript-tutorial/cs.javascript.info: 1-js/02-first-steps/16-function-expressions/article.md

# Funkční výrazy

Funkce v JavaScriptu není nějaká „magická jazyková struktura“, ale speciální druh hodnoty.

Syntaxe, kterou jsme dosud používali, se nazývá *deklarace funkce*:

```js
function řekniAhoj() {
  alert( "Ahoj" );
}
```

Existuje i jiná syntaxe pro vytváření funkce, která se nazývá *funkční výraz*.

Ten nám umožňuje vytvořit novou funkci uprostřed libovolného výrazu.

Například:

```js
let řekniAhoj = function() {
  alert( "Ahoj" );
};
```

> Zde je funkce vytvořena a výslovně přiřazena do proměnné jako kterákoli jiná hodnota. Nezáleží na tom, jak je funkce definována. Je to jen hodnota uložená do proměnné `řekniAhoj`.

Reviewer: ```suggestion
Zde vidíme proměnnou `řekniAhoj`, do níž je uložena hodnota, kterou je nová funkce, vytvořená jako `function() { alert("Ahoj"); }`.
```

> Význam obou částí kódu je stejný: „vytvořit funkci a uložit ji do proměnné `řekniAhoj`“.

Reviewer: ```suggestion
Protože k vytvoření funkce dochází v kontextu přiřazovacího výrazu (na pravé straně `=`), je to *funkční výraz*.
```

## Funkce je hodnota

Zopakujme si: bez ohledu na to, jak je funkce vytvořena, funkce je hodnota. V obou výše uvedených příkladech je funkce uložena do proměnné `řekniAhoj`.

Tuto hodnotu dokonce můžeme zobrazit pomocí `alert`:

```js run
function řekniAhoj() {
  alert( "Ahoj" );
}

*!*
alert( řekniAhoj ); // zobrazí kód funkce
*/!*
```

Všimněte si, že poslední řádek nevolá funkci, protože za `řekniAhoj` nejsou závorky. Existují programovací jazyky, v nichž je funkce zavolána pokaždé, když je uveden její název, ale to není případ JavaScriptu.

V JavaScriptu je funkce hodnota, takže s ní můžeme zacházet jako s hodnotou. Výše uvedený kód zobrazí její řetězcovou reprezentaci, kterou je zdrojový kód funkce.

Samozřejmě, funkce je speciální hodnota v tom smyslu, že ji můžeme volat, např. `řekniAhoj()`. Je to však stále hodnota, a proto s ní můžeme pracovat jako s hodnotou jakéhokoli jiného druhu.

Můžeme zkopírovat funkci do jiné proměnné:

```js run no-beautify
function řekniAhoj() {   // (1) vytvoření
  alert( "Ahoj" );
}

let funkce = řekniAhoj;  // (2) zkopírování

funkce(); // Ahoj        // (3) spustíme kopii (funguje to!)

> řekniAhoj(); // Ahoj     //     tohle pořád funguje taky (proč by nemělo)

Reviewer: ```suggestion
řekniAhoj(); // Ahoj     //     i tohle stále funguje (proč by nemělo)
```

1. Deklarace funkce `(1)` vytvoří funkci a uloží ji do proměnné nazvané `řekniAhoj`.
2. Řádek `(2)` ji zkopíruje do proměnné `funkce`. Opět si všimněte, že za `řekniAhoj` nejsou závorky. Kdyby tam byly, příkaz `funkce = řekniAhoj()` by zapsal do proměnné `funkce` *výsledek volání* `řekniAhoj`, ne samotnou *funkci* `řekniAhoj`.
3. Nyní můžeme tuto funkci volat pomocí `řekniAhoj()` i pomocí `funkce()`.

Na prvním řádku jsme mohli deklarovat `řekniAhoj` i pomocí funkčního výrazu:

```js
let řekniAhoj = function() { // (1) vytvoření
  alert( "Ahoj" );
};

let funkce = řekniAhoj;
// ...
```

Všechno bude fungovat stejně.

````smart header="Proč je na konci středník?"
Možná se divíte, proč funkční výraz obsahuje na konci středník `;`, ale deklarace funkce ne:

```js
function řekniAhoj() {
  // ...
}

let řekniAhoj = function() {
  // ...
}*!*;*/!*
```

Odpověď je jednoduchá: funkční výraz je zde vytvořen jako `function(…) {…}` uvnitř příkazu přiřazení: `let řekniAhoj = …;`. Středník `;` se doporučuje uvádět na konci příkazu, není to součást syntaxe funkce.

Středník by zde byl i u jednoduššího přiřazení, například `let řekniAhoj = 5;`, a je zde také u přiřazení funkce.
````

> ## Callbackové funkce

Reviewer: ```suggestion
## Funkce zpětného volaní (callback funkce)
```

## Funkční výrazy vs. deklarace funkcí

Zformulujme nyní základní rozdíly mezi deklaracemi funkcí a funkčními výrazy.

První rozdíl je syntaxe: jak je rozlišovat v kódu.

- *Deklarace funkce:* funkce deklarovaná samostatným příkazem v hlavním toku kódu:

    ```js
    // deklarace funkce
    function součet(a, b) {
      return a + b;
    }
    ```
- *Funkční výraz:* funkce vytvořená uprostřed výrazu nebo uvnitř jiného syntaktického konstruktu. Zde je funkce vytvořena na pravé straně „operátoru přiřazení“ `=`:

    ```js
    // funkční výraz
    let součet = function(a, b) {
      return a + b;
    };
    ```

Jemnější rozdíl spočívá v tom, *kdy* přesně je funkce vytvořena enginem JavaScriptu.

**Funkční výraz je vytvořen ve chvíli, kdy k němu dospěje vykonávání kódu, a je použitelný až od této chvíle.**

Jakmile se provádění skriptu dostane k pravé straně přiřazení `let součet = function…` -- v tom okamžiku je funkce vytvořena a od nynějška může být používána (přiřazena, volána apod.).

Deklarace funkcí jsou odlišné.

**Deklarovaná funkce může být volána ještě dříve, než je definována.**

Například globální deklarace funkce je viditelná v celém skriptu, ať se nachází kdekoli.

> Je to způsobeno vnitřními algoritmy. Když JavaScript připravuje spuštění skriptu, nejprve v něm najde globální deklarace funkcí a tyto funkce vytvoří. Můžeme to považovat za „inicializační krok“.

Reviewer: ```suggestion
Je to způsobeno vnitřními algoritmy. Když JavaScript připravuje spuštění skriptu, nejprve v něm najde globální deklarace funkcí a tyto funkce vytvoří. Můžeme to považovat za „přípravnou fázi“.
```

Například si představme, že potřebujeme deklarovat funkci `uvítání()`, která závisí na proměnné `věk`, kterou získáme za běhu skriptu. A pak ji plánujeme použít o něco později.

Kdybychom použili deklaraci funkce, nefungovalo by to tak, jak zamýšlíme:

```js run
let věk = prompt("Kolik je vám let?", 18);

// podmíněná deklarace funkce
if (věk < 18) {

  function uvítání() {
    alert("Ahoj!");
  }

} else {

  function uvítání() {
    alert("Dobrý den!");
  }

}

// ...použijeme ji později
*!*
uvítání(); // Chyba: uvítání není definováno
*/!*
```

Je to proto, že deklarace funkce je viditelná jedině uvnitř bloku kódu, v němž se nachází.

Další příklad:

```js run
let věk = 16; // vezmeme 16 jako příklad

if (věk < 18) {
*!*
  uvítání();               // \   (spustí se)
*/!*
                           //  |
  function uvítání() {     //  |  
    alert("Ahoj!");        //  |  deklarace funkce je k dispozici
  }                        //  |  všude v bloku, v němž je funkce deklarována
                           //  |
*!*
  uvítání();               // /   (spustí se)
*/!*

} else {

  function uvítání() {    

>     alert("Zdravíme vás!");

Reviewer: ```suggestion
    alert("Dobrý den!");
```
